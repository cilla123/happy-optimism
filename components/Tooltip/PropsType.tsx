export default interface PropsType {
  visible?: boolean;
  message?: any;
  children?: any;
}

declare global {
  interface Window {
    zarmTooltip?: any;
  }
}
