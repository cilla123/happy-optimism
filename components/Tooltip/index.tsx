import React, { PureComponent, cloneElement } from 'react';
import ReactDOM, { findDOMNode } from 'react-dom';
import classnames from 'classnames';
import PropsType from './PropsType';
// import Events from '../utils/events';

export interface TooltipProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Tooltip extends PureComponent<TooltipProps, any> {

  static defaultProps = {
    prefixCls: 'za-tooltip',
    visible: false,
  };

  private child;

  componentDidMount() {
    if (!window.zarmTooltip) {
      window.zarmTooltip = document.createElement('div');
      document.body.appendChild(window.zarmTooltip);
    }
    this.show(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.show(nextProps);
  }

  show = (props) => {
    const { prefixCls, className, visible, message } = props;
    const cls = classnames(prefixCls, className, {
      [`${prefixCls}-hidden`]: !visible,
    });

    // eslint-disable-next-line
    const rect = findDOMNode(this.child).getBoundingClientRect();
    const scrollTop = document.documentElement.scrollTop + document.body.scrollTop;
    const style = {
      left: rect.left,
      top: rect.top + scrollTop,
      width: rect.width,
    };

    ReactDOM.render(
      <div className={cls} style={style}>
        <div className={`${prefixCls}-inner`}>{message}</div>
      </div>
      , window.zarmTooltip);
  }

  render() {
    const { children } = this.props;
    return cloneElement(children, {
      ref: (ele) => { this.child = ele; },
    });
  }
}
