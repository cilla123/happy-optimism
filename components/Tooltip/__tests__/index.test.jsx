import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tooltip from '../index';

describe('Tooltip', () => {
  it('renders correctly', () => {
    const wrapper = render(<Tooltip message="hello"><span>foo</span></Tooltip>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('visible', () => {
    jest.useFakeTimers();
    const wrapper = mount(<Tooltip visible message="hello"><span>foo</span></Tooltip>);
    expect(toJson(wrapper)).toMatchSnapshot();
    jest.runAllTimers();
    wrapper.unmount();
  });
});
