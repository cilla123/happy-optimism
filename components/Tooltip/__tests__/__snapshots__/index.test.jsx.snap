// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders correctly 1`] = `
<span>
  foo
</span>
`;

exports[`Tooltip visible 1`] = `
<Tooltip
  message="hello"
  prefixCls="za-tooltip"
  visible={true}
>
  <span>
    foo
  </span>
</Tooltip>
`;
